### DISCUSSION TITLE
Restricting access to trpc

### DISCUSSION THREAD

**Discussion by karlbateman** [trpc/trpc#2383] [Q&A] [answered]
Hi all,

I have a NextJS application using tRPC (bootstrapped with the fantastic [create-t3-app](https://github.com/t3-oss/create-t3-app)), what is the process for restricting access to TRPC such that it can only be called by my NextJS client?

**Comment by KATT** [comment, marked as answer]
https://trpc.io/docs/authorization

**Reply by karlbateman** [reply]
Thanks for the link @KATT and for TRPC, it's awesome!

I'll provide more context as I don't think my original question was very clear. I have a contact form which is using TRPC to issue an API call to an SMTP service provider. What I'm hoping to achieve is to restrict access such that these requests can only be issued by my NextJS application. To be clear, this page is public and doesn't require a user to be authenticated.

Apologies if I've missed something obvious but I couldn't find any documentation for restricting TRPC calls to a specific client and/or host.

**Reply by KATT** [reply]
If it's public and not protected, what is the motivation of trying to "protect" this?

**Reply by KATT** [reply]
I guess what you're looking for is CSRF tokens? I don't have an example to pull out for you as I'm on my phone, but it should follow the same principle in trpc as elsewhere

**Reply by karlbateman** [reply]
If it's publicy accessible it can be targeted by scripts, bots etc. CSRF tokens is an option but doesn't feel like the right approach as the server doesn't maintain client state (and therefor no user sessions).

I could use Recaptcha to protect the form but I'd hoped there'd be a way to restrict access more generally. Maybe this is something which is best done elsewhere, such as a Cloudflare Page Rule.

**Reply by KATT** [reply]
I'm out of ideas, maybe some server generated token that can be validated on entry simply by an algorithm and then can only be used once? Essentially CSRF, doesn't need state apart from storing used keys
